### ISSUE TITLE
shall we have a Particle class

### ISSUE THREAD

**Issue by davydden** [dealii/dealii#3080]
In many cases one uses particles together with FEM. I think we should introduce a few classes to help manage particles and their relationship to a FE mesh. I can quickly draft a PR, but i would first like to discuss general ideas/options.

Following [Apect's particle](https://github.com/geodynamics/aspect/blob/master/include/aspect/particle/particle.h), I think something basic like

```
namespace particles 
{
#ifdef DEAL_II_WITH_64BIT_INDICES
         typedef unsigned long long int particle_index;
#else
         typedef unsigned int particle_index;
#endif

    template <int spacedim>
    class PointMassParticle
    {
         Point<spacedim> current_location;
         Point<spacedim> reference_location;
         particle_index  id;

        // plus get/set and default/non-default constructors
    }
}
```

would be good to have in deal.II. I called it `PointMassParticle` to make it clear that the particle has no dimension, i.e. it is a point-mass particle not a sphere or a circle, which would also have some geometric attributes.

When one associates particles to cells, there are different needs. For applications I am interested in, i need the relationship both ways: each particle should know about a cell it belongs, each cell should know a list of particles that belong to it.

In Aspect, an auxiliary class [Word](https://github.com/geodynamics/aspect/blob/master/include/aspect/particle/world.h) is introduced to manage particles. Particles are stored as a `std::multimap` from cell (level/index) to particle. Unfortunately, this would not be flexible enough for general usage when a user may need to store/use an association like `particle I -> particle J, K, L` to compute something.

So I guess we have to leave the management of particles to users. For the case above one could organize particles as

```
std::map<particles::particle_index, Particle<dim>> all_particles;
std::map<typename DoFHandler<dim>::active_cell_iterator, some-data-to-store-ids-of-particles> cells_to_particles;
```

Nonetheless, I think there should still be the recommended way to associate particles to cells. To that end one could have something like

```
namespace particles 
{
    template <int spacedim>
    class ParticleInCell : public PointMassParticle
    {
         Point<spacedim> location_within_cell; // quadrature point
         unsigned int active_cell_index; // or something else like Cell ID ?
         // plus a few get-set functions
    }
}
```

which would extend a general particle to the one which is associated to a cell. Given that one may be using MPI and `p::d::Tria` and may need to re-assign particles to cells owned by another processor, it's probably best to use `CellId` to cover all cases.

Do you think such classes would be useful to have inside the deal.II library? If so, how particles should relate to the cells they are in? Shall we leave management (i.e Aspect's Word) to users or have some class to do this together with cell -to - particle list association?

@bangerth @tjhei ping.
@jppelteret : I think you have some particles in the elasto-magneto case, not sure if those are point-mass particles or real spherical particles in the domain?
@vishalkenchan : further comments/ideas from our application/needs?

**Comment by vishalkenchan** [comment]
I am a fairly new user of dealii and still not at a level to fully grasp it's intricacies. Thank you @dealiidevelopers.

The application I and @davydden have in mind is one in which particles interact with each other within a certain cutoff radius. This would mean that a particle belonging to a cell could potentially interact with a particle inside a cell three or more cells away depending on cutoff radius, `p::s::Tria` and metis/parmetis partitioning of `p::s::Tria`.

Considering that each particle is an instantiation of the class PointMassParticle, I worry about data locality. Had I stored the particles (Point<dim>) in a container (say Vector), I might have a better data locality. The number of particles that could be loaded without considerable L1/L2/L3 cache misses would be low if each particle is an instantiation of the class. May be this is not really an issue or is it when the number of particles are around some millions or more? @tjhei any comments?

One way is to have such a map with particle and cell associations. The other way is to make particles a much more fundamental attribute of cells, just as the way DoFs are. In the latter case, perhaps, instead of a vector storing `global_dof_index`, we would be storing `Point<dim>`. And more importantly ParticleHandler (ParticleObjects and so on) could inherit from base class which is a minimal version of DoFHandler (DOFObjects and so on). I can understand that this would require significant programming effort. Looking forward to this discussion.

**Comment by gassmoeller** [comment]
Thanks for bringing up this discussion. When we developed the particle implementation in ASPECT we had for some time in mind that this funtionality might move to deal.II at some point and tried to keep it general. So I am happy to help implementing something new or moving some of those parts to deal.II.

I agree that there are certain parts that could definitely be provided by deal.II, among those are:
- A particle definition as you provided above  (similar to the `Particle` class of ASPECT)
- In my opinion we should also provide a manager class, e.g. `ParticleHandler` (similar to the `World` class of ASPECT). The reason we should provide such a class is that some of the tasks of particle-in-cell methods are common to most problems, and still relatively complicated to implement in a scalable way. One has to think of storing and loading particles during mesh refinement or serialization; communicating particles across subdomains in distributed memory calculations; also we spent significant effort in our implementation on keeping particle methods scalable despite using adaptive meshes (with vastly different numbers of particles per cell). 
- We should provide a way to output particles either with the DataOut class, or with something similar. Currently in ASPECT we wrote the output ourselves, because DataOut does not support writing point properties (patches are always expected to be at least lines as far as I know).

Then there are some parts that might or might not be provided by deal.II. That could be the following:
- In nearly all particle application the particles move in some way. Therefore it might be convenient to hand over a velocity solution (or tell the ParticleHandler which component of a finite element contains the velocity and hand over a DOFHandler and solution vector) and let deal.II handle the particle advection (in ASPECT we implemented different advection schemes, but we might also just implement a simple scheme, and let users expand if they want to).
- Particle generation can be everything from very simple (just create a number of particles in random positions) to very complicated (creating regular patterns, using probability density distributions over the domain). Should we provide at least a "default" way to fill the domain (represented by the triangulation) with particles?

Things that definitely should stay in user code are:
- Definitions of particular properties that are attached to particles. I do not know which properties your particles will represent, but ours often represent a certain chemical composition, or the amount of accumulated strain, therefore we need some flexible properties. Their number of components is not known at compile-time, but constant over the runtime of the model (although the values of the components might change during the model run).

To your discussion points @davydden: 
If the particles are stored in a map (or at least a CellID and ParticleID are stored in a map, if you worry about data locality and prefer to have them in separate vectors) you have quite a good association both ways. In my experience most often you do not have one object of type `Particle` and want to do something with it, but rather you have a task that should be done on all particles, and then one can iterate over the map, and each iterator contains a cell reference and a particle reference. Via the cell reference you can also reach neighbouring particles, e.g. by asking the map for all particles in neighbouring cells. Most tasks we do in ASPECT look like the following:

```
for cell in cells
  for particle in particle_map.find_all_particles(cell)
    do_something_with_particle
```

This way you can also use the fact that you can do every task that applies to a cell *once per cell* instead of *once per particle in this cell*, this is in particular valuable if you want to evaluate solutions or derivatives of solutions on the cell.

I agree that a common alternative approach is to use doubly-linked lists between cells and particles, and there you can do similar optimizations, but I never compared the two approaches in practice. The map seems to perform good enough for our purposes so far. We had models with approximately as many particles as degrees of freedom, and tested the scaling up to approximately fifty million degrees of freedom and particles (and some coarse tests suggest it continues to more than a billion particles / DoFs).

@vishalkenchan: You have a good point about data locality. It is not optimal in ASPECT's current design. However, storing them in a vector is not a perfect solution as well, because they will move around to different cells, and unless you sort your vector every few timesteps it might be as bad as in the map. In my experience doing everything *per cell* and accepting a not perfect data locality for the particles is better than doing everything *per particle* and doing the same work for a cell *n_particles_in_cell* times. But of course that might be specific to our application (where the particles strongly interact with the finite-element solution, but do not interact with each other).

**Comment by bangerth** [comment]
I have little to add to what @gassmoeller already nicely summarized. I think the implementation he has come up with in ASPECT is very good, and that there would be great value in looking there before re-inventing the wheel.

I definitely think that some functionality could and maybe should move to deal.II over time. I imagine that in particular (i) the Particle class itself, and (ii) a class that manages a collection of particles could move. The latter would, in essence, contain the `std::multimap<cell,Particle>` we currently have, plus a few algorithms that move particles from one cell to another, and from one processor to another.

I think that things that go beyond that will require a bit more thought.

I've always thought that before moving code it would be useful to first deal with one aspect I'm currently not too fond of: right now, every particle manages its own memory where it stores properties. This is wasteful -- we'd be better off with a central pool, and I have an idea how to design a class for this that I'd be happy to share if anyone was interested in implementing this.

**Comment by jppelteret** [comment]
@davydden Thanks, but no this doesn't apply to me. I have "real" particles/inclusions rather than point particles.

**Comment by davydden** [comment]
@gassmoeller : thanks for the detailed writeup. I agree with most of what you say, especially those things related to the Particle class and a class to manage them and do some basic output.

> The reason we should provide such a class is that some of the tasks of particle-in-cell methods are common to most problems, and still relatively complicated to implement in a scalable way.

:+1:


@bangerth :
> I imagine that in particular (i) the Particle class itself, and (ii) a class that manages a collection of particles could move.

Myself and @vishalkenchan are interested in those two.

My only concern with the current Aspect's implementation of the particle manager is that it does not provide a random access to a single particle within a cell. This is certainly needed for the case of interacting particles where one would have a double loop: for all particles within a cell, for all particles that interact with a given particle (could be just a single particle from a neighbouring cell).

I would suggest something like

```
std::map<types::LevelInd, std::vector<Particle<dim>>>
```
This way one could also iterate over all particles attributed to a cell, IMHO without any penaltisies compared to the `std::multimap<types::LevelInd, Particle<dim>>`. What is important is that one will also have random access to a single particle within a cell.


> we'd be better off with a central pool, and I have an idea how to design a class for this that I'd be happy to share if anyone was interested in implementing this.

I am interested in adding the very basic functionality for particle and a class to manage them to deal.II.  Shall I create a WIP PR with `Particle` and `ParticleManager` classes so that we can discuss there how to do a central pool of memory for particles?

**Comment by bangerth** [comment]
@davydden -- you can iterate over all elements in a `multimap` that correspond to the same key. That's because the range `particle_multimap.begin()` to `particle_multimap.end()` is sorted, and you can get a sub-range that corresponds to the same key. We do this all the time in ASPECT.

As for the memory pool for particles: I've written up my ideas here: https://github.com/geodynamics/aspect/issues/1215 . I think it would probably be easiest to implement this in ASPECT first (so it can be tested and, possibly, benchmarked) and then move the finished class to deal.II once we're happy with the result. This way we can avoid the divergence of functionality between two codes.

**Comment by davydden** [comment]
@bangerth :

> you can iterate over all elements in a multimap that correspond to the same key

I know that, but it is not enough. For interacting particles one need a **random access** to a particle  associated to a cell. That is, one needs an access of the form: give me a single n-th particle of the some cell. As far as I can tell, the following container satisfies needs of interacting particles:
```
std::map<types::LevelInd, std::vector<Particle<dim>>>
```
Would it be anyhow limiting to ASPECT needs? From iterators point of view, it should be pretty much the same as `std::multimap<types::LevelInd, Particle<dim>>`.

**Comment by bangerth** [comment]
That data type clearly satisfies these requirements. (But has implications for memory layout that are probably difficult to predict theoretically.) I suspect it would actually be quite simple to try this out for ASPECT, without too many code changes. But I don't think I quite understand why you actually need this. If you had `[b,e)` be the begin and end of the range of particles on one cell, and you want the `n`th, why isn't `b+n` not good enough? It's true that that is `O(n)`, but `n` is probably not very large.

Can you explain what exactly you need this for?

**Comment by davydden** [comment]
>  If you had [b,e) be the begin and end of the range of particles on one cell, and you want the nth, why isn't b+n not good enough? It's true that that is O(n), but n is probably not very large.

correct, that is also an option to get to n-th particle within a cell. Must admit i did not though about it before.

> Can you explain what exactly you need this for?

For interacting particles you always have a double loop:
```
for all particles in a cell
    for all neihbours of a particle  // <--- not needed in ASPECT
         do something // i.e. calculate interaction / integrate Newton equations, etc
```
The interaction is often limited to a certain raidius. Therefore, depending the (i) mesh size, (ii) interaction radius and (iii) number of particles, all the following situations are possible:
* a single particle interacts only with other particles inside the same cell
* a single particle interacts with very few particles from neighbouring cells (i.e. 2-3 particles out of 100)
* a single particle interacts with all particles within the same cell, and all particles within a few layers of neighbouring cells. 


For our case, the number of particles within a cell ranges from few (1-2) to about few hudreed ( @vishalkenchan would you agree with this upper bound?) for very heretogeneous mesh with big difference in element sizes.


To sum up, i agree that `std::multimap` will also do with `iterator+n`. Not sure how large `n` should be to see the difference in real-live application between `O(n)` in `std::multimap` and `O(1)` for `std::map<...,std::vector<>>`.

**Comment by bangerth** [comment]
So from your pseudocode, there you even need to iterate over *all* particles within a cell (both within the first as well as the second loop). Because there is no natural sorting of particles within cells, you can't even up front say that you only want a subset of the particles in a cell without explicitly testing each particle (e.g., for distance less than something to another particle). In that case, there really isn't any reason why you can't use the existing data structure -- just write the existing two loops in terms of iterators into the multimap with the same key. In other words, you don't even need `iterator+n`, you just start with `iterator=b` and then do `++iterator` in both the inner and outer loops.

**Comment by davydden** [comment]
> you can't even up front say that you only want a subset of the particles in a cell without explicitly testing each particle

these things are normally done separately using some neighbour lists_ to save time testing and also save time iterating over neighbours which actually don't have any effect on a given particle within a cell. Those neighbour lists are updated from time to time (say every 5-10 iterations). That's the reason the second loop should not be done via  `++iterator` and needs random access.

**Comment by bangerth** [comment]
I see. But in that case would you save any time if you stored particles on every cell as a vector, rather than as a subset of `multimap` elements? I mean, in the first case you'd store a set of indices into the neighbor's vector; in the latter case, you'd store a set of multimap iterators to identify a subset of the neighbor's particles.

I really don't have any empirical evidence that one approach would be better or faster than the other. I'm just trying to figure out your requirements to see why you want one data structure over another.

**Comment by davydden** [comment]
> But in that case would you save any time if you stored particles on every cell as a vector, rather than as a subset of multimap elements? I mean, in the first case you'd store a set of indices into the neighbor's vector; in the latter case, you'd store a set of multimap iterators to identify a subset of the neighbor's particles.

Right, both data structure can be used in neighbour lists. 

> I really don't have any empirical evidence that one approach would be better or faster than the other. 

neither do i

> why you want one data structure over another.

I did not originally though about `iterator+n` access for `std::multimap`. So at this point in discussion, I don't really want one over another. The `std::map`+`std::vector` was the first thing that came to mind knowing that I need access to a subset of particles. Thinking of it more, `std::multimap` may even be advantageous for the second loop... Need to think about it more.

**Comment by vishalkenchan** [comment]
> would you agree with this upper bound?
I would have to guess but in benchmark problems I have seen some cells would be associated to a few thousand (16K) particles.
> But has implications for memory layout that are probably difficult to predict theoretically.
I suppose this argument supports use of `std::map<types::LevelInd, Particle<dim>>`. While discussing with @davydden, I also didn't think through `iterator+n` logic and that push_bash would potentially resize `std::vector<Particle<dim>>` (size of this vector is not known apriori).

**Comment by bangerth** [comment]
That last argument was exactly my concern. We add and remove (from the middle) particles on cells all the time. This implies potentially having to reallocate memory frequently, leaving already allocated memory unused, and having to copy particles around if you delete one from the middle of one of the cells' vector.

**Comment by davydden** [comment]
@bangerth @gassmoeller :   me and @vishalkenchan were discussing whether or not `Particle<dim>`-like type are `std::is_pod<T>::value` and `std::is_trivial<T>::value` and it looks like this won't be the case, specifically at least due to aggregation of `Point<dim>`. I was curious what such particle types shall be so that they are guaranteed to have good memory alignment in `std::multimap<CellIterator, Particle<dim>>`?

**Comment by bangerth** [comment]
Yes, I think the design of the ASPECT particle class is pretty close to what I would propose we eventually move over to deal.II.

Regarding particle properties: Yes, these have run-time length, and so we wanted to keep them outside the particle itself. There is an open github issue for ASPECT to not use `std::vector<double>` (or some raw pointer) but to use a memory pool because the memory pool could allocate one big chunk of memory and split it into small parts of equal size. This would be more efficient, but we found that it requires using allocators that have state and that wasn't allowed pre-C++11.

**Comment by davydden** [comment]
closing this as we have a Particle class https://github.com/dealii/dealii/pull/4729 and Particle Handler is a WIP https://github.com/dealii/dealii/pull/5125
